### ISSUE TITLE
icalrestrction.c: VEVENT -> ICAL_RESTRICTION_ZEROORONE or ICAL_RESTRICTION_ONE

### ISSUE THREAD

**Issue by dilyanpalauzov** [libical/libical#345]
My `libical/src/libical/icalrestrction.c` contains:
```
    {ICAL_METHOD_NONE, ICAL_VCALENDAR_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZEROORONE, NULL},
    {ICAL_METHOD_NONE, ICAL_VEVENT_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZEROORONE, NULL},
    {ICAL_METHOD_NONE, ICAL_VTODO_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZEROORONE, NULL},
    {ICAL_METHOD_NONE, ICAL_VJOURNAL_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZEROORONE, NULL},
    {ICAL_METHOD_NONE, ICAL_VFREEBUSY_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ONE, NULL},
    {ICAL_METHOD_NONE, ICAL_VTIMEZONE_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZERO, NULL},
    {ICAL_METHOD_NONE, ICAL_XSTANDARD_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZERO, NULL},
    {ICAL_METHOD_NONE, ICAL_XDAYLIGHT_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZERO, NULL},
    {ICAL_METHOD_NONE, ICAL_XAUDIOALARM_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZERO, NULL},
    {ICAL_METHOD_NONE, ICAL_XDISPLAYALARM_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZERO, NULL},
    {ICAL_METHOD_NONE, ICAL_XEMAILALARM_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZERO, NULL},
    {ICAL_METHOD_NONE, ICAL_XPROCEDUREALARM_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ZERO, NULL},
    {ICAL_METHOD_NONE, ICAL_VAVAILABILITY_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ONE, NULL},
    {ICAL_METHOD_NONE, ICAL_XAVAILABLE_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ONE, NULL},
    {ICAL_METHOD_NONE, ICAL_VPOLL_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ONE, NULL},
    {ICAL_METHOD_NONE, ICAL_VPATCH_COMPONENT, ICAL_UID_PROPERTY, ICAL_RESTRICTION_ONE, NULL},
```

According to https://tools.ietf.org/html/rfc5545#section-3.6 for a VEVENT, VTODO, VJOURNAL, VFREEBUSY the UID is required but must not occur more than once, hoever libical validates a VEVENT without UID, see https://github.com/cyrusimap/cyrus-imapd/issues/2346 .

**Comment by ksmurchison** [comment]
My fear in fixing this is that it will break existing software that is exploiting/depending on this incorrect behavior.  

@winterz  What do you think?

**Comment by ksmurchison** [comment]
Obviously, if we fix this it would be done in design-data/restrictions.csv

**Comment by ksmurchison** [comment]
Sure

**Comment by ksmurchison** [comment]
@winterz  I started auditing restrictions.csv against 5545 and 5546.  UID and DTSTAMP were both optional in 2445.  Do we still want to support 2445/2446?  Those specs also supported EXRULE and multiple RRULE.

**Comment by winterz** [comment]
so that explains it.
I should have remembered that.

yeah, I think we still need to support 2445/2446, do you?

**Comment by ksmurchison** [comment]
As much as I hate it, I'm guessing we probably have to (even though 5545 is 8+ years old).  That said, I wonder how many pieces of software actually create "real" components with no UID.  In order for scheduling to work properly, a UID is required.

I'm still going to audit restrictions.csv and fix/reorganize it.  I might even add a restriction that if a component includes ORGANIZER (scheduling object) that it MUST include UID.

**Comment by ksuther** [comment]
@ksmurchison I'm not sure what you mean by "real" components, but I've seen a few web-based ics generators which don't include UIDs (usually things that create a tiny ics for an appointment reminder). That's not part of a scheduling system or anything, but when it hits libical on our end we always check to make sure a UID exists and then create one if it doesn't.

**Comment by ksmurchison** [comment]
I used "real" in the libical sense: VEVENT, VTODO, VJOURNAL, as opposed to VALARM, VTIMEZONE

**Comment by winterz** [comment]
I guess I'll put something in the documentation that RFC5545/46 is not 100% validated as we soften requirements on UID in order to support older RFCs.

I just know that within seconds of taking away rfc2245/46 support we'll get yelled at.

**Comment by winterz** [comment]
interestingly:
the ReadMe.txt does not mention rfc2245/46.
the doc/Mainpage.dox does not mention rfc5545/46

seems to me that previous teams working on the source code have also had some trouble deciding how to handle this situation.

at least in 3.0.x series I'll try to make the documentation consistent.

**Comment by winterz** [comment]
hmm... @ksmurchison could you make a patch that forces rfc5545?
I'm thinking we should try that out in master and see what happens.

**Comment by ksmurchison** [comment]
sure

**Comment by winterz** [comment]
revisit this for 3.1 release

**Comment by winterz** [comment]
revisit this for the 4.0 release
